image/layout/tar.go

```diff
@@ -0,0 +1,236 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package layout
+
+import (
+	"archive/tar"
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"time"
+
+	"github.com/opencontainers/image-spec/specs-go"
+	"golang.org/x/net/context"
+)
+
+// TarEntryByName walks a tarball pointed to by reader, finds an
+// entry matching the given name, and returns the header and reader
+// for that entry.  Returns os.ErrNotExist if the path is not found.
+func TarEntryByName(ctx context.Context, reader io.ReadSeeker, name string) (header *tar.Header, tarReader *tar.Reader, err error) {
+	_, err = reader.Seek(0, os.SEEK_SET)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	tarReader = tar.NewReader(reader)
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, nil, ctx.Err()
+		default:
+		}
+
+		header, err := tarReader.Next()
+		if err == io.EOF {
+			return nil, nil, os.ErrNotExist
+		}
+		if err != nil {
+			return nil, nil, err
+		}
+
+		if header.Name == name {
+			return header, tarReader, nil
+		}
+	}
+}
+
+// WriteTarEntryByName reads content from reader into an entry at name
+// in the tarball at file, replacing a previous entry with that name
+// (if any).  The current implementation avoids writing a temporary
+// file to disk, but risks leaving a corrupted tarball if the program
+// crashes mid-write.
+//
+// To add an entry to a tarball (with Go's interface) you need to know
+// the size ahead of time.  If you set the size argument,
+// WriteTarEntryByName will use that size in the entry header (and
+// Go's implementation will check to make sure it matches the length
+// of content read from reader).  If unset, WriteTarEntryByName will
+// copy reader into a local buffer, measure its size, and then write
+// the entry header and content.
+func WriteTarEntryByName(ctx context.Context, file io.ReadWriteSeeker, name string, reader io.Reader, size *int64) (err error) {
+	var buffer bytes.Buffer
+	tarWriter := tar.NewWriter(&buffer)
+
+	_, err = file.Seek(0, os.SEEK_SET)
+	if err != nil {
+		return err
+	}
+
+	tarReader := tar.NewReader(file)
+	found := false
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
+
+		var header *tar.Header
+		header, err = tarReader.Next()
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			return err
+		}
+
+		if header.Name == name {
+			found = true
+			err = writeTarEntry(ctx, tarWriter, name, reader, size)
```

Comment:
On Tue, Aug 30, 2016 at 01:44:20PM -0700, Stephen Day wrote:
> > +			err = writeTarEntry(ctx, tarWriter, name, reader, size)
> 
> I don't get the point of this model if it is only useful for a
> file-backed engine.

WriteTarEntryByName and writeTarEntry are helpers that are only useful
for the tar-backed engines (and we have CAS and ref engines backed by
tar files).  Zip-backed engines (which I'll add if/when this PR lands)
will probably need similar helpers.  Directory-backed engines (which
I'll add if/when this PR lands unless folks want them to be part of
this PR) will probably not need similar helpers.

> Is there an engine that supports a buffer and batch model that
> doesn't require memory to be allocated for all resources?

You can avoid buffering the tar file in memory if you write the new
tar file to disk and copy it over the original path, replacing the
file descriptor you were using with one that references the new file.
Any other engines which were using file descriptors referencing the
old instance will not notice the change.

I'm ok with doing that, but it means we'll probably want to adjust the
TarEngine functions to store the path (where they currently store the
ReadWriteSeekCloser) and have them open (and possibly
write/rename-over) the file for every operation.  That's doable, but
the current implementation is sufficient for testing, so I haven't
bothered.  If you'd rather me take that (or another) approach in this
PR, just let me know.